### PR TITLE
refactor(compiler): shorten _nghost- and _ngcontent- prefixes to _h- and _c-

### PR DESCRIPTION
**Pull request by ngwattcos**
This replaces all occurrences of _nghost- to _h- and _ngcontent to _c-, including in docs and tests.

Shortening _nghost and _ngcontent will significantly reduce the overhead from adding additional attributes to HTML elements.

This allows for smaller bundles, which can improve download speeds and reduce parsing time.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

**Comment by jelbourn** [comment]
@ngwattcos looks like there are still some outstanding test failures:
* `integration-payloads.json` looks like it needs to be updated
* On the angular/components repo, we have a test for `MatBadge` in `badge.spec.ts` that tests for the presence of `ngcontent` to verify that ViewEncapsulation is applied; that test will need to be updated before this PR could go in
* There seems be be a test in `legacy-unit-tests-saucelabs` that is looking at the attribute names

**Comment by ngwattcos** [comment]
@jelbourn Made a pull request to badge.spec.ts. However, I'm not clear on what needs to be changed in ```integration-payloads.json```. Also, could you point me to the test on ```legacy-unit-tests-saucelabs```? Thanks.

**Comment by jelbourn** [comment]
@ngwattcos you should be able to find the specific test failures from the CircleCI result pages. 

@pkozlowski-opensource @alxhub FYI this change is from a Google summer intern and was proposed by Igor. What are your thoughts on changing these encapsulation attributes? Shortening the string seems like a good idea to me, but I'm not totally sure whether we would want to do this in a minor or a major. In angular/components, we had one unit tests that was looking at this attribute. Inside Google, there are 11 places where `_ngcontent` is referenced and 4 for `_nghost`, mostly for people doing an exact HTML comparison.

**Comment by pkozlowski-opensource** [comment]
tl;dr; I like the direction of the change but IMO it is breaking enough so that we can do it only in major.

> FYI this change is from a Google summer intern and was proposed by Igor. What are your thoughts on changing these encapsulation attributes? Shortening the string seems like a good idea to me, but I'm not totally sure whether we would want to do this in a minor or a major. In angular/components, we had one unit tests that was looking at this attribute. Inside Google, there are 11 places where _ngcontent is referenced and 4 for _nghost, mostly for people doing an exact HTML comparison.

Just to make sure that I've got proper context here - my understanding is that we are trying to reduce size of the generated code for styles, right? As such the idea sounds really good (still would be interesting to see real-life savings data as those common prefixes should compress quite well?).

The change in the code looks good but it reveals how many tests depend on those prefixes in the repo - I would expect this to be the case in the wild as well. In fact we can see that even some high-profile projects use  _ngcontent / _nghost in their tests: https://github.com/search?l=TypeScript&q=_nghost&type=Code

Apart from tests I'm also worried that people do hacks in their code, mostly to work around various issues - example from a random GitHub search: https://github.com/alexurquhart/ng-tippy/blob/6b4109aae5d4b569971b8f67d638339e18a864d0/src/app/tooltip.directive.ts#L41-L49

Given all this I would say that this is a braking change (in the sense that it will break people projects, even if this isn't formally our public API). As such I would do it only in major and consider mitigation strategies (ex. automatic migration or  configuration option). 

I will do more in-depth review of this PR when it gets green on CI.

**Comment by jelbourn** [comment]
@pkozlowski-opensource that's some good evidence, I'm now convinced that this should only happen in a major version

@ngwattcos @kseamon this would mean that realistically we wouldn't be able to merge this change until we're on the 11.0.0 release train, which is at least a few months away (I don't know the exact timeframe off the top of my head)

**Comment by petebacondarwin** [comment]
@ngwattcos - could you rebase this PR? It has conflicts. Unfortunately we missed the 11.0.0 release, so now this PR will need to target the 12.0.0 release in the spring.

**Comment by jelbourn** [comment]
For context: the author of this PR was a Google summer intern. Since their internship is long over, this is something we'd have to pick up ourselves if we want to move forward with it.

**Comment by petebacondarwin** [comment]
In the context of https://github.com/angular/angular/issues/25355 I would like to consider taking on this PR but changing it so that we add `data-` to the front. I know that this will undue some of the benefit from shortening the names. But if we are going to go to the hassle of updating these then we might as well make the attributes W3C valid.

cc @jelbourn

**Comment by petebacondarwin** [comment]
Closing this for similar reasons to that given for closing https://github.com/angular/angular/pull/40595#issuecomment-820561403.
